### ISSUE TITLE
[feature] Merged options in transitive dependencies

### ISSUE THREAD

**Issue by rddesmond** [conan-io/conan#10283]
I've been working with the awesome experimental components feature and how to build exactly what you need.  My particular goal is analyzing security impacts of my project, and one of the strategies is minimizing the vulnerability surface.  The useful feature would be to merge options when a dependency is required by multiple 

```
Root -> A, B
A -> C (with feature X)
B -> C (don't need feature X)
```

The result of this is that you need to decide that you need feature X in Root.  C is no longer a simple transitive dependency, but one you have to specify deliberately.  It would be useful if there could be a hook for logic C to say something like `X requested = any(X needed)`.

For components specifically, there could also be more complex logic.  What if there was an option to declare the required components?

```
Root -> A, B
A -> C (components = X,Y)
B -> C (components = Y,Z)
```

Then the logic would be `X requested = ','.join(set(X needed.split(',')))`.  This might not be the best for every use case (because it would result in package for every combination of components), but it would allow a simple way to only build exactly what you want.  This is particularly important if the components aren't quite independent (what if, Z only works if X has a compile-time reference to it?).

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @rddesmond 

I think what you are looking for is the ``cpp_info[component].requires`` feature: https://docs.conan.io/en/latest/creating_packages/package_information.html#using-components

You can see there something like:

```python
def package_info(self):
        self.cpp_info.components["comp1"].requires = ["zlib::zlib"]  # Depends on all components in zlib package
        self.cpp_info.components["comp2"].requires = ["comp1", "openssl::ssl"]
```

That explicitly defines that you will only be requiring ``openssl::ssl`` component but not the others. As long as the build system and the build system integration supports this (at least CMakeDeps does), I think this is the behavior you want.

**Comment by rddesmond** [comment]
@memsharded That was what I was jumping off of!  It's great, but this feature is one step more.  Instead of changing the build of your package, this would have it effect the build of the dependency, too.  So, have it dynamically determine the minimum build of the dependency which has options and a `packageId` that would satisfy the needs of the dependencies.

In my first example, package `C` is required transitively twice.  One use requires feature X and one doesn't.  This feature would be so the value of feature X is determined automatically to be `ON`, without having to resort to setting it higher up.  (Similarly, if neither use required feature Y, it isn't built.)

The goal would be to allow a minimal build of `C`, not just minimal use of it.  It would allow the user to monitor everything that is built, reduces the vulnerability surface, and makes it easier to remediate if you have a SAST scanner that is watching the build of everything.

**Comment by memsharded** [comment]
Oh, I think I see now.

The problem with this is that it seems it is not possible, it would be a chicken and egg problem. Lets say that we have package B-(depends on)->package A.

- if the usage from downstream consumer B can imply a different ``package_id()`` of A, it means the evaluation of ``package_info()`` of the consumer B, needs to happen before the package A ``package_id()`` is computed.
- The ``package_info()`` of every package cannot be evaluated until the binary is built and packaged, because it depends on the actual binaries being built.
- In order to be able to execute the ``package_info()`` of consumer B, it should have been built from source, which means having computed its ``package_id``. But package B ``package_id`` depends on its dependency A, so package A full reference, including its ``package_id`` needs to be known to compute package B one.

It seems the consumption/component-requirement definition cannot be coupled with the binary model. The only possibility is the one that is already implemented, direct definition of what is built in every package via options, like ``with_feature=True/False`` that is already done in many recipes in ConanCenter.

**Comment by memsharded** [comment]
Hi @rddesmond 

Closing this as unplanned/not possible.

The behavior in Conan 2 has now changed, and it is no longer encouraged to change options values in recipes: https://docs.conan.io/2/knowledge/faq.html#defining-options-for-dependencies-in-conanfile-py-recipes-doesn-t-work, and in general defining options values in profile is much more recommended, and it will more easily define exactly which components to define in your whole product, consistently across builds.

Please don't hesitate to open new tickets as necessary for any further question or issue. Thanks very much for your feedback.!
